### ISSUE TITLE
fireEvent with Promises

### ISSUE THREAD

**Issue by timdeschryver** [testing-library/eslint-plugin-testing-library#166]
Based on https://github.com/testing-library/dom-testing-library/issues/609, I think we could add a new rule. I'm not 100% sure if it's worth it, because:

- testing library already added a runtime error
- in TypeScript projects, this will already be caught (I think)
- the current implementation of the rule doesn't catch all cases, and I'm not sure if it can. Currently it catches `fireEvent(Promise)`, `fireEevent(new Promise()`, `fireEvent(findBy...)`

The implementation of the rule might look like this.
What do you tink?

```ts
export default ESLintUtils.RuleCreator(getDocsUrl)<Options, MessageIds>({
  name: RULE_NAME,
  meta: {
    type: 'suggestion',
    docs: {
      description: 'TK',
      category: 'Best Practices',
      recommended: false,
    },
    messages: {
      noPromiseInEvent: 'TK',
    },

    fixable: 'code',
    schema: [],
  },
  defaultOptions: [],

  create(context) {
    return {
      'ImportDeclaration[source.value=/testing-library/]'(
        node: TSESTree.ImportDeclaration
      ) {
        const fireEventImportNode = node.specifiers.find(
          specifier =>
            isImportSpecifier(specifier) &&
            specifier.imported &&
            'fireEvent' === specifier.imported.name
        ) as TSESTree.ImportSpecifier;

        const {references} = context.getDeclaredVariables(fireEventImportNode)[0];

        for (const reference of references) {
          const referenceNode = reference.identifier;
          if (
            isMemberExpression(referenceNode.parent) &&
            isCallExpression(referenceNode.parent.parent)
          ) {
            const [element] = referenceNode.parent.parent
              .arguments as TSESTree.Node[];
            if (element) {
              if (isCallExpression(element) || isNewExpression(element)) {
                const methodName = isIdentifier(element.callee)
                  ? element.callee.name
                  : isMemberExpression(element.callee) &&
                    isIdentifier(element.callee.property)
                  ? element.callee.property.name
                  : '';

                if (
                  ASYNC_QUERIES_VARIANTS.includes(methodName) ||
                  methodName === 'Promise'
                ) {
                  context.report({
                    node: element,
                    messageId: 'noPromiseInEvent',
                  });
                }
              }
            }
          }
        }
      },
    };
  },
});
```

**Comment by Belco90** [comment]
Oh wow, I didn't even know about this error and I just realized I suffered it!

It could be a good addition to help users spotting the issue in this particular case, though having already a runtime error from testing library seems enough, doesn't it? Not sure what we should do, but you almost got the implementation 😅

**Comment by gndelia** [comment]
> in TypeScript projects, this will already be caught (I think)

confirming that ☝️ 

![image](https://user-images.githubusercontent.com/352474/84788098-1d1bca00-afc5-11ea-8467-0a0908b2d404.png)


The library (in runtime) and TS (in compile time) also offer a good coverage against the scenario, but given the implementation is almost done, I'd say let's go with it 😄

**Comment by timdeschryver** [comment]
👍 I'll add it later this week with and add test cases/docs, we can still decide if we want to add it then 🙂
Thank god I added the code here.... since I just reverted by changes locally 😂

**Comment by Belco90** [comment]
Sorry I had to do it
![image](https://user-images.githubusercontent.com/2677072/84993655-dc998900-b149-11ea-8a3b-36e22445836f.png)
